### PR TITLE
zarf package mirror command introduction

### PR DESCRIPTION
**Pull request by Racer159**
## Description

This adds a command that mirrors artifacts from a package and pushes them to their airgap equivalents

## Related Issue

Fixes #N/A

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging

- [ ] Test, docs, adr added or updated as needed
- [X] [Contributor Guide Steps](https://github.com/defenseunicorns/zarf/blob/main/CONTRIBUTING.md#developer-workflow) followed

### CODE REVIEW

File: src/pkg/packager/unload.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	"github.com/defenseunicorns/zarf/src/config"
+	"github.com/defenseunicorns/zarf/src/pkg/message"
+	"github.com/defenseunicorns/zarf/src/pkg/utils"
+	"github.com/defenseunicorns/zarf/src/pkg/utils/helpers"
+	"github.com/defenseunicorns/zarf/src/types"
+)
+
+// Unload pulls resources from a package (images, git repositories, etc) and pushes them to remotes in the air gap without deploying them
+func (p *Packager) Unload(packageName string) (err error) {
+	spinner := message.NewProgressSpinner("Unloading zarf package %s", packageName)
```
Comment (YrrepNoj) [inline src/pkg/packager/unload.go]: Unsure how this is intended to be called, but we should probably use the package name from the `p *Packager` metadata.

Comment (Racer159) [reply src/pkg/packager/unload.go]: yep yep - refactored

**Comment by Madeline-UX** [comment]
@Racer159 Created a test for naming of new command. Results of the test will appear on the [Maze Results page]( https://app.maze.co/report/New-Feature-Naming-Text/7fpilkd04gok/intro) Planning to get participants early next week -

**Comment by laurentiusoica** [comment]
Are there plans to include mirroring for helm charts as well? And also configuration to control the git repository names when mirroring?

File: src/cmd/package.go [hunk 2]
```diff
@@ -102,6 +103,28 @@ var packageDeployCmd = &cobra.Command{
...
+		pkgConfig.PkgSource = pkgConfig.PkgOpts.PackagePath
+
+		// Configure the packager
+		pkgClient := packager.NewOrDie(&pkgConfig)
+		defer pkgClient.ClearTempPaths()
+
+		// Deploy the package
+		if err := pkgClient.Mirror(noImgChecksum); err != nil {
+			message.Fatalf(err, lang.CmdPackageDeployErr, err.Error())
+		}
```
Comment (Noxsios) [inline src/cmd/package.go:124]: For consistency `noImgChecksum` should be made into `MirrorOpts`, as all other top level packager receivers do not take arguments and instead grab from the opts struct.

File: src/cmd/package.go [hunk 3]
```diff
@@ -334,6 +359,27 @@ func bindDeployFlags(v *viper.Viper) {
...
 
+func bindMirrorFlags(v *viper.Viper) {
+	mirrorFlags := packageMirrorCmd.Flags()
+
+	// Always require confirm flag (no viper)
+	mirrorFlags.BoolVar(&config.CommonOptions.Confirm, "confirm", false, lang.CmdPackageDeployFlagConfirm)
+
+	mirrorFlags.BoolVar(&noImgChecksum, "no-img-checksum", false, lang.CmdPackageMirrorFlagNoChecksum)
+
+	mirrorFlags.StringVar(&pkgConfig.PkgOpts.OptionalComponents, "components", v.GetString(common.VInitComponents), lang.CmdPackageMirrorFlagComponents)
```
Comment (Noxsios) [inline src/cmd/package.go:370]: Mirror resources by default uses `v.GetString(common.VInitComponents)`?

File: src/pkg/packager/mirror.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+	"github.com/defenseunicorns/zarf/src/config"
+	"github.com/defenseunicorns/zarf/src/pkg/message"
+	"github.com/defenseunicorns/zarf/src/pkg/utils/helpers"
+	"github.com/defenseunicorns/zarf/src/types"
+)
+
+// Mirror pulls resources from a package (images, git repositories, etc) and pushes them to remotes in the air gap without deploying them
+func (p *Packager) Mirror(noImgChecksum bool) (err error) {
+	spinner := message.NewProgressSpinner("Mirroring zarf package %s", p.cfg.PkgOpts.PackagePath)
```
Comment (Noxsios) [inline src/pkg/packager/mirror.go:19]:
```suggestion
	spinner := message.NewProgressSpinner("Mirroring Zarf package %s", p.cfg.PkgOpts.PackagePath)
```

File: src/pkg/packager/mirror.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+			}
+		}
+	}
+
+	return nil
+}
+
+// mirrorComponent mirrors a Zarf Component.
+func (p *Packager) mirrorComponent(component types.ZarfComponent, noImgChecksum bool) error {
+	message.Debugf("packager.mirrorComponent(%#v, %#v", p.tmp, component)
```
Comment (Noxsios) [inline src/pkg/packager/mirror.go:66]:
```suggestion
```

File: src/pkg/packager/mirror.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+		if err := p.pushImagesToRegistry(component.Images, noImgChecksum); err != nil {
+			return fmt.Errorf("unable to push images to the registry: %w", err)
+		}
+	}
+
+	if hasRepos {
+		if err = p.pushReposToRepository(componentPath.Repos, component.Repos); err != nil {
+			return fmt.Errorf("unable to push the repos to the repository: %w", err)
+		}
+	}
```
Comment (Noxsios) [inline src/pkg/packager/mirror.go:89]: this is where I would put debug statements if I wanted them, as all of the previous operations in this fn are near guaranteed to not fail.

**Review by Noxsios — CHANGES_REQUESTED** [review]

File: src/cmd/package.go (hunk 3 above)
Comment (Racer159) [reply src/cmd/package.go:370]: Le Copy Le Pasta

File: src/cmd/package.go (hunk 2 above)
Comment (Racer159) [reply src/cmd/package.go:124]: Yeah that was from back when that wasn't the case - will refactor

File: src/config/config.go [hunk 7]
```diff
@@ -66,6 +66,7 @@ const (
 
 	ZarfDeployStage = "Deploy"
 	ZarfCreateStage = "Create"
+	ZarfUnloadStage = "Unload"
```
[the remark this answers is not in the document]
Comment (Racer159) [reply src/config/config.go]: The way these are used the actual text of the string is printed and would need to be resolved (so an enum that is an integer would require a secondary slice to decode it).

File: src/pkg/packager/mirror.go (hunk 6 above)
Comment (Racer159) [reply src/pkg/packager/mirror.go]: I think it is fine to not add debug statements - removed them from the other places (was mostly copy pasta at that point)

**Review by cmwylie19 — APPROVED** [review]
lgtm

**Review by Noxsios — APPROVED** [review]

**Pull request merged**